src/utils.ts

```diff
@@ -68,21 +62,53 @@ type VerifyJWTOptions = {
 }
 
 export async function verifyJWT({ jwt, secretKey }: VerifyJWTOptions) {
-  try {
-    const secret = new TextEncoder().encode(secretKey)
-    const { payload } = await jwtVerify(jwt, secret)
-    return payload
-  } catch (err: unknown) {
-    throw new Error(ERRORS.INVALID_JWT)
-  }
+  const secret = new TextEncoder().encode(secretKey)
+  const { payload } = await jwtVerify<TOTPPayload>(jwt, secret)
+  return payload
 }
 
 /**
  * Miscellaneous.
  */
-export function ensureStringOrUndefined(value: unknown) {
+
+export function coerceToOptionalString(value: unknown) {
   if (typeof value !== 'string' && value !== undefined) {
     throw new Error('Value must be a string or undefined.')
   }
   return value
 }
+
+export function coerceToOptionalNonEmptyString(value: unknown) {
+  if (typeof value === 'string' && value.length > 0) return value
+  return undefined
+}
+
+export function coerceToOptionalTotpData(value: unknown) {
+  if (
+    typeof value === 'object' &&
+    value !== null &&
+    'hash' in value &&
+    typeof (value as { hash: unknown }).hash === 'string' &&
+    'attempts' in value &&
+    typeof (value as { attempts: unknown }).attempts === 'number'
+  ) {
+    return value as TOTPData
+  }
+  return undefined
+}
+
+export type RequiredAuthenticateOptions = Required<
+  Pick<AuthenticateOptions, 'failureRedirect' | 'successRedirect'>
+> &
+  Omit<AuthenticateOptions, 'failureRedirect' | 'successRedirect'>
+
+export function assertIsRequiredAuthenticateOptions(
+  options: AuthenticateOptions,
+): asserts options is RequiredAuthenticateOptions {
+  if (options.successRedirect === undefined) {
+    throw new Error(ERRORS.REQUIRED_SUCCESS_REDIRECT_URL)
+  }
+  if (options.failureRedirect === undefined) {
+    throw new Error(ERRORS.REQUIRED_FAILURE_REDIRECT_URL)
+  }
```

Comment:
Interesting, I would love to hear @mw10013's opinion on this, as he currently has control over the strategy, although I'm the one who has the last words.

> Implementing this escape hatch would delegate the responsibility of managing the sessionTotpKey cleanup to the developer.

- This sounds like a risk I'm not sure we should take. Yes, it's on the developer's side, but this could open gaps in the implementation and probably lead to people complaining to us about the strategy not working _(due to bad developer implementation)_.

I could be wrong about this, since I've not checked The Epic Stack links, and instead am just sharing some lazy thoughts. Would love @mw10013's opinion on this.

Feel free to open a PR, or keep discussing this with us @mechanicalbot!

Reply:
@mechanicalbot: your comment is very interesting and giving me food for thought. Would love to hear more about what you are thinking and your particular user cases.

I agree with you that `failureRedirect` and I can't say I'm enamored of the approach using redirects for control flow. You are right that we are using `failureRedirect` to ensure our updates to the session vars are committed. With `Error` (ie. following `throwOnError: true`) instead of redirect, I can't think of a way to ensure the session vars are updated and committed consistently. Am I overlooking something here?

The Epic Stack authentication flow is more sophisticated and complex. From what I remember from the design notes, TOTP is used to authenticate the user's email and then the user can create an account with username and password.

`remix-auth-totp` has a simpler flow and goals

- Authenticate an email using TOTP.
- Ask application for `user` corresponding to the authenticated email.
- Simple integration with no database or worries about cleaning up expired TOTP's

Please say more about a generic "auth callback" to manage users and session independently. What particular use case are you facing that would benefit from this? Would there be a way for `remix-auth-totp` to manage the session yet allow the application to manage the users independently?

Reply:
@mechanicalbot: Thanks for the additional color. Can we walk step-by-step through a scenario so I can better understand more concretely.

- User is at a page that has links to login by TOTP or external providers like google and facebook.
- User clicks on TOTP `/login' link.
- User submits email on `/login' form.
- `remix-auth-totp` stores the email in session, generates the TOTP code, and asks application to `sendTOTP` to the user.
- User clicks on magic link in the email.
- `remix-auth-totp` validates the code, gets the `user` from the application by calling `verify`, stores `user` in session using the session key provided by `remix-auth`, and finally throws `successRedirect`.

Can you layout this scenario in the same step-by-step way, but include what modifications and changes you would like to see to match your use case.

Reply:
@mechanicalbot: Indeed, pseudo code > step-by-step outline > verbiage

```ts
// authenticator.server.ts
authenticator.use(
  new GitHubStrategy(
    {
      // ...
    },
    async ({ profile }) => {
      const email = profile.emails[0].value;
      return { email };
    },
  ),
);

authenticator.use(
  new TOTPStrategy(
    {
      // ...
    },
    async ({ email }) => {
      return { email };
    },
  ),
);

// magic-link.tsx
export async function action({ request }: DataFunctionArgs) {
  await authenticator.authenticate("TOTP", request, {
    successRedirect: "/auth/totp/callback", // currently required by remix-auth-totp cause session vars
    failureRedirect: "/login",
  })
}

// auth.$provider.callback.tsx
export async function loader({ request }: DataFunctionArgs) {
    const session = await sessionStorage.getSession(request.headers.get('cookie'))
    const profile = session.get(authenticator.sessionKey) ?? null // Need to somehow get the authenticator or its sessionKey

  const email = profile.email;
    let user = await db.user.findFirst({ where: { email } })
    if (!user) {
      // Mutation in GET (I hang my head in shame).
      user = await db.user.create({ data: { email } })
    }
  
    return createSessionCookieAndRedirectSomehow(user, "/account");
}

export async function action({ params, request }: DataFunctionArgs) {
  const profile = await authenticator.authenticate(params.provider, request, {
    failureRedirect: "/login",
  })

  const email = profile.email;
  let user = await db.user.findFirst({ where: { email } })
  if (!user) {
    user = await db.user.create({ data: { email } })
  }
  
  return createSessionCookieAndRedirectSomehow(user, "/account");
}```

Reply:
@mechanicalbot: Clever. Here's my attempt based on your upgrade to my understanding.

```ts
// authenticator.server.ts
authenticator.use(
  new GitHubStrategy(
    {
      // ...
    },
    async ({ profile }) => {
      const email = profile.emails[0].value;
      return { email };
    },
  ),
);

authenticator.use(
  new TOTPStrategy(
    {
      magicLinkPath: "/auth/totp/callback",
      // ...
    },
    async ({ email }) => {
      return { email };
    },
  ),
);

// auth.$provider.callback.tsx
export async function loader({ request }: DataFunctionArgs) {
  try {
    await authenticator.authenticate(params.provider, request, {
      successRedirect: "/SWEET_SMELL_OF_SUCCESS",
      failureRedirect: "/login",
  } catch (throwable) {
    if (throwable instanceof Response && new URL(throwable.url).pathname === "/SWEET_SMELL_OF_SUCCESS") {
      const session = await sessionStorage.getSession(throwable.headers.get('set-cookie'))
      const profile = session.get(authenticator.sessionKey) ?? null;
      if (!profile) throw new Error("Reversal of fortune.");

      const email = profile.email;
      let user = await db.user.findFirst({ where: { email } })
      if (!user) {
        user = await db.user.create({ data: { email } })
      }
      return redirect("/off-to-the-races", {
        headers: { // preserve session cause remix-auth-totp
          'set-cookie': await sessionStorage.commitSession(session, {
            maxAge: MAX_AGE, 
          }),
        },
      })
    }
    throw throwable;
  }
```